src/backend/utils/adt/orderedsetaggs.c

```diff
@@ -1165,7 +1168,7 @@ mode_final(PG_FUNCTION_ARGS)
 		CHECK_FOR_INTERRUPTS();
 	}
 
-	if (shouldfree && !last_val_is_mode)
+	if (shouldfree && !last_val_is_mode && last_val)
```

Comment:
I think if `last_val` is 0 it means its value is not valid and should not be used even as an argument for `DatumGetPointer` and checking for `DatumGetPointer(last_val)` will only obscure the code.

Reply:
> I think if `last_val` is 0 it means its value is not valid

This might be not true generally. As `last_val` has polymorphic `Datum` type and that it's a reference is defined via `osastate->qstate->typByVal` flag, zero might be valid value for "passing by value" case.

IMO the note from @InnerLife0 might be useful for code readability: the fact that we work with reference value in `last_val` variable by previously verifying `shouldfree` flag. I'd add around `DatumGetPointer(last_val)` the macro `PointerIsValid()` to make this even more readable.

Reply:
Ok, thank you for explanation, seems to be reasonable. Fixed. Thank you for comment.